src/ray/gcs/tables.cc

```diff
@@ -2,101 +2,147 @@
 
-Status ClientTable::Connect(ClientID *client_id) {
-  auto data = std::make_shared<ClientTableDataT>();
-  data->client_id = client_id_.binary();
-  // TODO(swang): Get the address and port from somewhere.
-  data->node_manager_address = "";
-  data->node_manager_port = 0;
+  auto data = std::make_shared<ClientTableDataT>(local_client_);
   data->is_insertion = true;
-
-  // TODO(swang):
-  // - Add ourselves to the client table (easier if this is synchronous).
-  // - Subscribe to the client table.
-  // - Once subscription is complete, read all client table entries once.
-
-  *client_id = client_id_;
-  return Status::OK();
+  // Callback for a notification from the client table.
+  auto notification_callback = [this](AsyncGcsClient *client,
+                                      const ClientID &channel_id,
+                                      std::shared_ptr<ClientTableDataT> data) {
+    return HandleNotification(client, channel_id, data);
+  };
+  // Callback to handle our own successful connection once we've added
+  // ourselves.
+  auto add_callback = [this](AsyncGcsClient *client, const ClientID &id,
+                             std::shared_ptr<ClientTableDataT> data) {
+    HandleConnected(client, id, data);
+  };
+  // Callback to add ourselves once we've successfully subscribed.
+  auto subscription_callback = [this, data, add_callback](
+      AsyncGcsClient *c, const ClientID &id, std::shared_ptr<ClientTableDataT> d) {
+    // Mark ourselves as deleted if we called Disconnect() since the last
+    // Connect() call.
+    if (disconnected_) {
+      data->is_insertion = false;
+    }
+    return Add(JobID::nil(), client_id_, data, add_callback);
+  };
+  return Subscribe(JobID::nil(), ClientID::nil(), notification_callback,
+                   subscription_callback);
 }
 
 Status ClientTable::Disconnect() {
-  auto data = std::make_shared<ClientTableDataT>();
-  data->client_id = client_id_.binary();
-  // TODO(swang): Get the address and port from somewhere.
-  data->node_manager_address = "";
-  data->node_manager_port = 0;
-  data->is_insertion = false;
-
-  // TODO(swang):
-  // - Add ourselves to the client table (easier if this is synchronous).
+  auto data = std::make_shared<ClientTableDataT>(local_client_);
+  data->is_insertion = true;
+  auto add_callback = [this](AsyncGcsClient *client, const ClientID &id,
+                             std::shared_ptr<ClientTableDataT> data) {
+    HandleConnected(client, id, data);
+  };
+  RAY_RETURN_NOT_OK(Add(JobID::nil(), client_id_, data, add_callback));
+  // We successfully added the deletion entry. Mark ourselves as disconnected.
+  disconnected_ = true;
   return Status::OK();
 }
 
+const ClientTableDataT &ClientTable::GetClient(const ClientID &client_id) {
+  RAY_CHECK(!client_id.is_nil());
+  auto entry = client_cache_.find(client_id);
+  if (entry != client_cache_.end()) {
+    return entry->second;
```

Comment:
Should this include a check for `entry->second.is_insertion`?

Reply:
No, it includes clients that have disconnected.